[RF Docs] Active Support Instrumentation (#16)
Ridhwana/rails: guides/source/active_support_instrumentation.md

core component of Ruby on Rails, offering a wide range of Ruby language
extensions and utility classes.

One of its key features is the Instrumentation API, which is described in
further detail in this guide. You will learn:

* About Instrumentation
* How to instrument your own events.
* How to subscribe to events.
* The instrumentation hooks available inside Rails.

--------------------------------------------------------------------------------

Introduction to Instrumentation
-------------------------------

The Instrumentation API provides a way to instrument code and subscribe to
events that occur in your application. Instrumentation means wrapping a block of
code so that, when it runs, an event is emitted with a name and optional
payload. Any subscribers listening for that event will then be notified and can
react, for example, they can log information, benchmark, or perform some other
action. This makes it possible to observe behavior within the Rails framework,
in your own code, or even in standalone Ruby scripts. There are a few parts that
are vital to understanding the Instrumentation API: hooks, events, and
subscribers.

> 

Reviewer: We should probably have a NOTE here to make clear that instrumentation should be used for framework code and not application code, and link to "Active Support Event Reporter" for that instead.

> ### Hooks

Reviewer: I'm wondering if "Hook" is basically the same as "Instrumentation" and using Hook only adds confusion.
[The Notifications API documentation](https://edgeapi.rubyonrails.org/classes/ActiveSupport/Notifications.html) doesn't mention "Hook" at all.

Author: You're right! I updated this to use the terms "Instrumentation" and "Events" instead, and removed the hooks section and all terminology.  

The most notable changes were:
- moving the hooks examples into the events section
- renaming Rails Framework Hooks to Rails Framework Instrumentation, which I think works well, but feel free to let me know if you think otherwise. 

https://github.com/Ridhwana/rails/pull/16/commits/8f2423c8c4fe006342c1310ccb6a563ba321ec4d

Hooks help us observe behavior within the Rails framework; these hooks are
predefined points in the framework where an event is emitted. By subscribing to
an event from a hook, you can run your own code whenever that event occurs.

For example, there is [a hook](#sql-active-record) that is called every time
Active Record executes a SQL query on a database. This hook can be subscribed

> to, and used to track the number of queries during a certain action. There's

Reviewer: This introduction section is pretty wordy. Maybe we can show example code for some of the examples instead?

````suggestion
For example, there is [a  `sql.active_record` hook](#sql-active-record) that is called every time
Active Record executes a SQL query on a database. This hook can be subscribed
to, and used to track the number of queries during a certain action.

```ruby
sql_counter = 0
ActiveSupport::Notifications.subscribe "sql.active_record" do |event|
   sql_counter += 1
end
```

There's
````

[another hook](#process-action-action-controller) which is called when
processing an action of a controller. This hook can be subscribed to, and used
to track how long a specific action has taken. You can read more about hooks in

> the [Rails framework hooks section](#rails-framework-hooks) later in this guide.

Reviewer: ```suggestion
[another hook](#process-action-action-controller) which is called when
processing an action of a controller. This hook can be subscribed to, and used
to track how long a specific action has taken.

```ruby
ActiveSupport::Notifications.subscribe "process_action.action_controller" do |event|
   Rails.logger.info "Duration: #{event.duration}" 
end
```

You can read more about hooks in
the [Rails framework hooks section](#rails-framework-hooks) later in this guide.
````

### Events

An event is generated when a hook is triggered or when you've [instrumented your
own event](#instrumenting-custom-events). It is a record of something that has

> happened. You can use events to track changes in your application.

Reviewer 2: Think it might be worth reversing and combining the first two sentences for a slightly better flow:

An event is a record of something that has happened, and is generated when a hook is triggered or when you've instrumented your own event.

Reviewer: I think we can clarify this a bit more by not distinguishing between Rails events and custom events.
"a hook is triggered" and "when you've instrumented your own event" basically describe the same thing, the first one is just already defined by Rails.

Maybe something like:
> An event is a record of something that has happened, and is generated when a instrumented block of code is called.

Or maybe:
> An event is a record of something that has happened, and is emitted when a instrumented block of code is called.

For example, when the
[`process_action.action_controller`](#process-action-action-controller) hook is
triggered, then an event is generated. This event has a name and optional data.
The name is `process_action.action_controller` and the data includes the
controller name, action name, and other information about the request.

You can read more about creating your own events and subscribing to events in
the  [Instrumenting Custom Events section](#instrumenting-custom-events) and
[Subscribing to an Event section](#subscribing-to-an-event) respectively.

### Subscribers

A subscriber is the object that is used to subscribe to events. It is used to
listen to events and perform some action when an event is emitted.

For example, if you want to subscribe to the `process_action.action_controller`
event and benchmark it, then you can create a subscriber that listens to that
event. The subscriber will then be able to perform other actions when the event
is emitted, such as logging the duration of the action.

A single event can have multiple subscribers. You can read more about
subscribers in the [Subscribing to an Event section](#subscribing-to-an-event).

> Instrumenting Custom Events

Reviewer 2: Just a thought: Do you think it could be better to have the section about 'Subscribing to an Event' next as it's maybe simpler/more common, and then have the 'Custom Events' section after that?

Reviewer: I think I agree with Harriet. Maybe even move it below the "Rails Framework Hooks", as it's probably more common to subscribe to built-in hooks than creating custom events/hooks? Maybe the custom hooks would be easily overlooked though.
I don't have a strong opinion on this.

Reviewer: Thinking about this some more, it does make sense to first describe "instrumenting events" before "subscribing to events", because it makes the "subscribing" section clearer. We can see how events are generated and what we are subscribing to.

Maybe this section title should put less emphasis on the "Custom" part and be renamed to just "Instrumenting Events". We could also add an introductional paragraph, that explains it's less common to create custom events and you typically use the built-in events, and that we show how to instrument events to make it clearer how things work.

Although I'm not sure if it's less common to create custom events than subscribing to built-in events, as I've never done either.

Author: > Thinking about this some more, it does make sense to first describe "instrumenting events" before "subscribing to events", because it makes the "subscribing" section clearer. We can see how events are generated and what we are subscribing to.

Thanks both, the above is the reason why I moved the sections around. Hence, I took this approach like Petrik suggested:

> Maybe this section title should put less emphasis on the "Custom" part and be renamed to just "Instrumenting Events". We could also add an introductional paragraph, that explains it's less common to create custom events and you typically use the built-in events, and that we show how to instrument events to make it clearer how things work.

---------------------------

You can instrument your own events, by calling
[`ActiveSupport::Notifications.instrument`](https://api.rubyonrails.org/classes/ActiveSupport/Notifications.html#method-c-instrument)
with the `name` of your custom event, a `payload` which is a hash containing
information about the event, and an optional block.

```ruby

> ActiveSupport::Notifications.instrument "my.custom.event", this: "payload" do

Reviewer: I've never added instrumentation to an app, but I think subscribing is commonly implemented in an initializer?
```suggestion
# config/initializers/instrumentation.rb
ActiveSupport::Notifications.instrument "my.custom.event", this: "payload" do
```

  # do your custom stuff here
end
```

TIP: You should follow Rails conventions when defining your own events. The
format is: `event.library`. For a blogging application, pick an emitter name
like posts (or blog if that’s your domain), then instrument names like

> `publish.posts`.

Reviewer 2: This tip confused me a little. The whole tip is referring to the instrument name right, i.e. what you have as 'my.custom.event' above? I was a bit lost on what an 'emitter name' was?

Author: Good catch here, I've improved it by moving things aroudn and rewording the tip section!

to subscribe to multiple events at once. Here's how to subscribe to everything
from `ActionController`:

```ruby
ActiveSupport::Notifications.subscribe(/action_controller/) do |**args|
  Rails.logger.info "#{args}"
end
```

In your Rails log, you'll see one line for each ActionController event. Example:

```bash
INFO -- : {:name=>"start_processing.action_controller", :id=>"a3f2e9...", :payload=>{:controller=>"PostsController", :action=>"index", :params=>{"controller"=>"posts", "action"=>"index"}, :format=>:html, :method=>"GET", :path=>"/posts"}}
INFO -- : {:name=>"process_action.action_controller", :id=>"a3f2e9...", :payload=>{:controller=>"PostsController", :action=>"index", :status=>200, :view_runtime=>12.34, :db_runtime=>3.21}}
```

### Subscribing to a Single Event using Multiple Subscribers

You can subscribe to a single event using multiple subscribers. This is useful
if you want to subscribe to an event from multiple sources.

For example, you may want to subscribe to a `publish.posts` event from multiple
sources to kick off a background job to send an email to the post author, and to
also log the event.

Rails Framework Hooks
---------------------

> Within the Ruby on Rails framework, there are a number of hooks provided for

Reviewer 2: NIT: worth adding 'already' here for emphasis?
```suggestion
Within the Ruby on Rails framework, there are a number of hooks already provided for
```

common events.

> Each event below lists the event name you can subscribe to, explains how the

Reviewer 2: Quite a lot of the word 'event' here which makes it a slightly more confusing read: 

```suggestion
Each heading below lists the event name you can subscribe to, explains how the
```

> event is triggered, and the corresponding example `event.payload` from the

Reviewer 2: I wonder if we just want payload here to make this a bit simpler to read as well. Happy to be ignored though!

```suggestion
event is triggered, and the corresponding example payload from the
```

subscribed event.

> To subscribe to a specific event, use

Reviewer 2: We have quite detailed support on how to subscribe to events in a previous section - do you think we could cut this part out and just link to the above?

Author: I initially found the Rails framework hooks a bit confusing because they didn’t explain how event subscription worked or how it connected to the tables shown under each header. Including the example helps make that relationship clearer, even if it adds a bit of length.

[`ActiveSupport::Notifications.subscribe`](https://api.rubyonrails.org/classes/ActiveSupport/Notifications.html#method-c-subscribe).
For example:

```ruby
ActiveSupport::Notifications.subscribe "process_action.action_controller" do |event|
  # Do something with the event
  Rails.logger.info event.name         # "process_action.action_controller"
  Rails.logger.info event.duration     # 10 (in milliseconds)
  Rails.logger.info event.allocations  # 1826
  Rails.logger.info event.payload

>   #   {

Reviewer: ```suggestion
  # {
```

  #   controller: "PostsController",
  #   action: "index",
  #   params: {"action" => "index", "controller" => "posts"},
  #   headers: #<ActionDispatch::Http::Headers:0x...>,
  #   format: :html,
  #   method: "GET",
  #   path: "/posts",
  #   request: #<ActionDispatch::Request:0x...>,
  #   response: #<ActionDispatch::Response:0x...>,
  #   status: 200,
  #   view_runtime: 46.848,
  #   db_runtime: 0.157
  # }
end
```

NOTE: `event.payload` is a hash with the payload of the event. Below we describe

> the keys, and examples values that you can expect to see in the payload.

Reviewer 2: ```suggestion
the keys that you can expect to see in the payload and give example values.
```

Read more about subscribing to the event in [Subscribing to an

> Event](#subscribing-to-an-event).

Reviewer 2: Ah - I see here you have just linked to the above as per my earlier comment. But still - perhaps it would be better if we could just reword this sentence a bit, move it earlier and lose the above example for brevity's sake.

### Action Cable

#### `perform_action.action_cable`

The event is emitted when a channel action is invoked.

For example:

```ruby
# Client sends: { "action":"speak", "message":"Hi" }
class ChatChannel < ApplicationCable::Channel
  def speak(data) = Message.create!(text: data["message"])
end
```

The event payload (`event.payload`) includes the following keys (with typical
example values).

> | Payload Key      | Description               | Example               |

Reviewer: If we rename the column header to "Example Value" (we also have "Payload Key"), the sentence can be simplified somewhat:
```suggestion
The event payload (`event.payload`) includes the following keys:

| Payload Key      | Description               | Example Value   |
```

Author: ✅ Thanks for this suggestion, it does make it a bit clearer and reduces the redundancy https://github.com/Ridhwana/rails/pull/16/commits/f03fe3adee1d69ca595c3b7ab85cd972a87c51ad

| ---------------- | ------------------------- | --------------------- |
| `:channel_class` | Name of the channel class | `"ChatChannel"`       |
| `:action`        | The action                | `"speak"`             |
| `:data`          | A hash of data            | `{ "message"=>"Hi" }` |

#### `transmit.action_cable`

The event is emitted when a message is transmitted from a channel.

For example:

```ruby
class ChatChannel < ApplicationCable::Channel
  def subscribed
    transmit(ok: true)
  end
end
```

The event payload (`event.payload`) includes the following keys (with typical
example values).

| Payload Key      | Description               | Example          |
| ---------------- | ------------------------- | ---------------- |
| `:channel_class` | Name of the channel class | `"ChatChannel"`  |
| `:data`          | A hash of data            | `{ "ok"=>true }` |
| `:via`           | Via                       | `"websocket"`    |

#### `transmit_subscription_confirmation.action_cable`

The event is emitted when a subscription confirmation is sent to a client.

For example:

```ruby
# Successful subscription auto-confirms -> event fires

> ```

Reviewer 2: The above code example is quite different from the others, which work a bit harder to try and emulate real functionality. Should we bring this one at least into line with the `transmit_subscription_rejection` example?

Author: Oops this was a mistake, updating with an example.

The event payload (`event.payload`) includes the following keys (with typical
example values).

| Payload Key      | Description               | Example         |
| ---------------- | ------------------------- | --------------- |
| `:channel_class` | Name of the channel class | `"ChatChannel"` |

#### `transmit_subscription_rejection.action_cable`

The event is emitted when a subscription rejection is sent to a client.

For example:

```ruby
def subscribed
  reject

> end

Reviewer: We already show in what class `subscribed` is defined in the example above, but I think it's clearer if we use it for other examples as well:
```suggestion
class ChatChannel < ApplicationCable::Channel
  def subscribed
    reject
  end
end
```

```

The event payload (`event.payload`) includes the following keys (with typical
example values).

| Payload Key      | Description               | Example         |
| ---------------- | ------------------------- | --------------- |
| `:channel_class` | Name of the channel class | `"ChatChannel"` |

#### `broadcast.action_cable`

The event is emitted when a broadcast is published to a stream.

For example:

```ruby
ActionCable.server.broadcast("chat_room_1", text: "Hello")
```

The event payload (`event.payload`) includes the following keys (with typical
example values).

| Payload Key     | Description        | Example                 |
| --------------- | ------------------ | ----------------------- |
| `:broadcasting` | Named broadcasting | `"chat_room_1"`         |

> | `:message`      | A hash of message  | `{ "text"=>"Hello" }`   |

Reviewer 2: ```suggestion
| `:message`      | A hash containing message data  | `{ "text"=>"Hello" }`   |
```

| `:method`       | HTTP request verb                                         | `"GET"`                                     |
| `:path`         | Request path                                              | `"/posts"`                                  |
| `:request`      | The [`ActionDispatch::Request`][] object                  | `#<ActionDispatch::Request ...>`            |
| `:response`     | The [`ActionDispatch::Response`][] object                 | `#<ActionDispatch::Response ...>`           |
| `:status`       | HTTP status code                                          | `200`                                       |
| `:view_runtime` | Amount spent in view in ms                                | `46.848`                                    |
| `:db_runtime`   | Amount spent executing database queries in ms             | `0.157`                                     |

#### `send_file.action_controller`

The event is emitted when a controller streams or sends a file with `send_file`.

For example:

```ruby
send_file "/var/app/exports/report.csv",
  filename: "report.csv",

>   disposition: "attachment"

Reviewer: I think this example would be clearer as well with a full example.
Something like:
```suggestion
class ReportsController < ApplicationController
  def download
    send_file "/var/app/exports/report.csv",
      filename: "report.csv",
      disposition: "attachment"
  end
end
```

```

The event payload (`event.payload`) includes the file path plus any options you
pass to send_file.

| Payload Key | Description               | Example                         |
| ----------- | ------------------------- | ------------------------------- |
| `:path`     | Complete path to the file | `"/var/app/exports/report.csv"` |

In this case, the payload will always contain the `:path` key, but will also
contain the `:filename`, and `:disposition` keys which are passed to the
`send_file` method.

#### `send_data.action_controller`

The event is emitted when a controller sends raw data with `send_data`.

For example:

```ruby
send_data "Hello, world!",
  filename: "report.txt",
  type: "text/plain",

>   disposition: "attachment"

Reviewer: ```suggestion
class ReportsController < ApplicationController
  def download
    send_data "Hello, world!",
      filename: "report.txt",
      type: "text/plain",
      disposition: "attachment"
  end
end
```

`redirect_to`.

For example:

```ruby
class PostsController < ApplicationController
  def old
    redirect_to posts_url
  end
end
```

The event payload (`event.payload`) includes the following keys (with typical
example values).

| Payload Key | Description                       | Example                             |
| ----------- | --------------------------------- | ----------------------------------- |
| `:status`   | HTTP response code                | `302`                               |
| `:location` | URL to redirect to                | `"http://localhost:3000/posts/new"` |
| `:request`  | The [`ActionDispatch::Request`][] | `#<ActionDispatch::Request ...>`    |

#### `halted_callback.action_controller`

The event is emitted when a before/around/after callback halts the action chain.

For example:

```ruby
class PostsController < ApplicationController
  before_action :require_login
end
```

The event payload (`event.payload`) includes the following keys (with typical
example values).

| Payload Key | Description                   | Example            |
| ----------- | ----------------------------- | ------------------ |

> | `:filter`   | Filter that halted the action | `":require_login"` |

Reviewer: Nitpick, but the usage of `filter` seems to be a remnant of controller callbacks using  `before_filter` instead of `before_action`. The payload key should probably be renamed as well in Rails.
```suggestion
| `:filter`   | Callback that halted the action | `":require_login"` |
```

The event is emitted when strong parameters filtering detects unpermitted keys.

For example:

```ruby
class UsersController < ApplicationController
  def create
    # params = { user: { name: "Ada", admin: true } }
    permitted = params.require(:user).permit(:name) # :admin is unpermitted
    User.create!(permitted)
  end
end
# => triggers unpermitted_parameters.action_controller
```

The event payload (`event.payload`) includes the following keys (with typical
example values).

| Payload Key | Description                                               | Example                                  |
| ----------- | --------------------------------------------------------- | ---------------------------------------- |
| `:keys`     | The unpermitted keys                                      | `["admin"]`                              |
| `:context`  | Hash with `:controller`, `:action`, `:params`, `:request` | `{ controller: "UsersController", ... }` |

#### `send_stream.action_controller`

The event is emitted when a controller streams data with `send_stream`.

For example:

```ruby
send_stream(filename: "subscribers.csv") do |stream|
  # ...

> end

Reviewer: ```suggestion
class SubscribersController < ApplicationController
  def download
    send_stream(filename: "subscribers.csv") do |stream|
      # ...
    end
  end
end
```

[`ActionMailbox::Base`]:
    https://api.rubyonrails.org/classes/ActionMailbox/Base.html

### Action Mailer

#### `deliver.action_mailer`

The event is emitted when a mail is delivered (after the message is generated
and delivery is attempted).

For example:

```ruby
UserMailer.welcome(current_user).deliver_now
```

The event payload (`event.payload`) includes the following keys (with typical
example values).

| Payload Key           | Description                   | Example                |
| --------------------- | ----------------------------- | ---------------------- |
| `:mailer`             | Name of the mailer class      | `"Notification"`       |
| `:message_id`         | ID of the message (Mail gem)  | `"<abc@host>"`         |
| `:subject`            | Subject of the mail           | `"Rails Guides"`       |
| `:to`                 | To address(es)                | `["users@rails.com"]`  |
| `:from`               | From address                  | `["me@rails.com"]`     |
| `:bcc`                | BCC addresses                 | `[]`                   |

> | `:cc`                 | CC addresses                  | `[]`                   |

Reviewer: "rails.com" is not related to Ruby on Rails, so maybe just use "example.com"?
We can also add examples for `:bcc` and `:cc`:
```suggestion
| `:to`                 | To address(es)                | `["users@example.com"]`  |
| `:from`               | From address                  | `["me@example.com"]`     |
| `:bcc`                | BCC addresses                 | `["bcc@example.com"]` |
| `:cc`                 | CC addresses                  | `["cc@example.com"]`  |
```

The ActionMailer::Base API docs use `"Order Watcher <watcher@example.com>"` but that might be a bit long?
https://edgeapi.rubyonrails.org/classes/ActionMailer/Base.html

> | `:date`               | Date of the mail              | `Sat, 10 Mar 2012 ...` |

Reviewer: Nitpick, but we might update the date here to be a bit more recent:
```suggestion
| `:date`               | Date of the mail              | `Mon, 10 Mar 2025 ...` |
```

> | `:mail`               | Encoded form of the mail      | `"(omitted)"`          |

Reviewer: Maybe remove the code fences here to make it a bit clearer it's not the actual result:
```suggestion
| `:mail`               | Encoded form of the mail      | Omitted                |
```

Or cut off the result similar to `:date`?
```suggestion
| `:mail`               | Encoded form of the mail      | `Date: Mon, 10 Mar 2025 ...`            |
```

| `:perform_deliveries` | Whether delivery is performed | `true`                 |

#### `process.action_mailer`

The event is emitted when a mailer action is invoked to build a message.

For example:

```ruby
UserMailer.welcome(current_user) # building the message triggers process.action_mailer
```

The event payload (`event.payload`) includes the following keys (with typical
example values).

| Payload Key | Description              | Example           |
| ----------- | ------------------------ | ----------------- |
| `:mailer`   | Name of the mailer class | `"Notification"`  |
| `:action`   | The action               | `"welcome_email"` |

> | `:args`     | The arguments            | `[]`              |

Reviewer: ```suggestion
| `:args`     | The arguments            | `[current_user]`  |
```

Or similar to other example values:
```suggestion
| `:args`     | The arguments            | `[#<User ...>]`  |
```

### Action View

#### `render_template.action_view`

The event is emitted when a full template (with optional layout) is rendered.

For example:

```ruby

> render :index, layout: "application"

Reviewer: It would be nice if we could show the locals from the example table here as well:
```suggestion
render :index, layout: "application", locals: { foo: "bar" }
```

```

The event payload (`event.payload`) includes the following keys (with typical
example values).

| Payload Key   | Description                        | Example                                |
| ------------- | ---------------------------------- | -------------------------------------- |
| `:identifier` | Full path to template              | `".../app/views/posts/index.html.erb"` |
| `:layout`     | Applicable layout                  | `"layouts/application"`                |
| `:locals`     | Local variables passed to template | `{ foo: "bar" }`                       |

#### `render_partial.action_view`

The event is emitted when a partial is rendered.

For example:

```erb

> <%= render "form", post: @post %>

Reviewer: Similarly as above, it would be nice if this examples matches with the table (another option would be using `post` for the table example.
```suggestion
<%= render "form", foo: "bar" %>
```

| Payload Key | Description                     | Example                 |
| ----------- | ------------------------------- | ----------------------- |
| `:job`      | Job object                      | `#<MyJob ...>`          |
| `:adapter`  | QueueAdapter processing the job | `#<...>`                |
| `:error`    | The error that caused the retry | `#<Timeout::Error ...>` |
| `:wait`     | The delay of the retry          | `30.seconds`            |

#### `enqueue_all.active_job`

The event is emitted when multiple jobs are enqueued together.

For example:

```ruby
jobs = [ MyJob.new, MyJob.new ]
ActiveJob::Base.enqueue_all(jobs) # when supported by the adapter
```

The event payload (`event.payload`) includes the following keys (with typical
example values).

| Payload Key | Description                     | Example                 |
| ----------- | ------------------------------- | ----------------------- |
| `:adapter`  | QueueAdapter processing the job | `#<...>`                |
| `:jobs`     | An array of Job objects         | `[ #<MyJob ...>, ... ]` |

#### `perform_start.active_job`

The event is emitted when job execution starts (on the worker).

For example:

```ruby

> # Fired on the worker right before perform begins

Reviewer: I think we can show `perform_now` here as `_perform_job` [emits it](https://github.com/rails/rails/blob/cf4e3c508adf930eb70e4b3e6e16d2ca83b7e9b3/activejob/lib/active_job/instrumentation.rb#L43).
```suggestion
MyJob.perform_now # Fired on the worker right before perform begins
```

```

The event payload (`event.payload`) includes the following keys (with typical
example values).

| Payload Key | Description                     | Example        |
| ----------- | ------------------------------- | -------------- |
| `:adapter`  | QueueAdapter processing the job | `#<...>`       |
| `:job`      | Job object                      | `#<MyJob ...>` |

#### `perform.active_job`

The event is emitted when job execution finishes (on the worker).

For example:

```ruby

> # Fired after perform finishes (success or handled failure)

Reviewer: ```suggestion
MyJob.perform_now
# Fired after perform finishes (success or handled failure)
```

| `:name`              | Name of the operation                            | `"Post Load"`                                       |
| `:binds`             | Bind parameters                                  | `[ #<ActiveModel::Attribute::WithCastValue ...> ]`  |
| `:type_casted_binds` | Typecasted bind parameters                       | `[11]`                                              |
| `:async`             | `true` if query is loaded asynchronously         | `false`                                             |
| `:allow_retry`       | `true` if the query can be automatically retried | `true`                                              |
| `:connection`        | Connection object                                | `#<ActiveRecord::ConnectionAdapters::...>`          |
| `:transaction`       | Current transaction, if any                      | `#<ActiveRecord::ConnectionAdapters::...>` or `nil` |
| `:affected_rows`     | Number of rows affected by the query             | `0`                                                 |
| `:row_count`         | Number of rows returned by the query             | `5`                                                 |

> | `:cached`            | `true` when result comes from the query cache    | `true`/`false`                                      |

Reviewer: Maybe just use a single example value here?
```suggestion
| `:cached`            | `true` when result comes from the query cache    | `false`                               |
```

| `:statement_name`    | SQL statement name (Postgres only)               | `nil`                                               |

#### `strict_loading_violation.active_record`

> The event is emitted when a lazily loaded association is accessed on a model

Reviewer 2: I would normally expect a hyphen here but I've got a vague feeling that both are technically acceptable, so feel free to overrule! 

```suggestion
The event is emitted when a lazily-loaded association is accessed on a model
```

| ------------- | ------------------------------------------------ | --------------------------------- |
| `:owner`      | Model with `strict_loading` enabled              | `#<User id: 1 ...>`               |
| `:reflection` | Reflection of the association that tried to load | `#<ActiveRecord::Reflection ...>` |

[`config.active_record.action_on_strict_loading_violation`]:
    configuring.html#config-active-record-action-on-strict-loading-violation

#### `instantiation.active_record`

The event is emitted when Active Record instantiates model objects from query
results.

For example:

```ruby
User.all.to_a
```

The event payload (`event.payload`) includes the following keys (with typical
example values).

| Payload Key     | Description                    | Example  |
| --------------- | ------------------------------ | -------- |
| `:record_count` | Number of records instantiated | `1`      |
| `:class_name`   | Record’s class                 | `"User"` |

#### `start_transaction.active_record`

The event is emitted when Active Record starts a database transaction (on first
DB interaction inside a `transaction` block).

> NOTE:  Active Record does not create the actual database transaction until

Reviewer: I'm not sure if we should maybe first show an regular example (with the payload table) before explaining the intricacies of when it gets started? This would make it more similar to the other examples but the actual moment the event is emitted might get overlooked. 🤷 
The note now also somewhat repeats the sentence above it.
````suggestion
```ruby
ActiveRecord::Base.transaction do
  User.count # Event fired here.
end
```

NOTE:  Active Record does not create the actual database transaction until
````

Author: I've moved it around a little bit, but kept the examples and explanations before the table, because it read a bit more naturally and felt consistent with other examples.

needed.

For example:

```ruby
ActiveRecord::Base.transaction do
  # We are inside the block, but no event has been triggered yet.

  # The following line makes Active Record start the transaction.

>   User.count  # Event fired here.

Reviewer: I think this extra space was unintended? Maybe also remove the period?
```suggestion
  User.count # Event fired here
```

```

However, if `requires_new: true` is passed, you get an event for the nested

> transaction too.

Reviewer 2: How about 

"an event is emitted for the nested transaction too."

?

| ------------- | ------------------------------------------ | ---------------------------------------- |
| `:reflection` | The reflection of the association          | `#<ActiveRecord::Reflection ...>`        |
| `:message`    | Descriptive message about the access       | `"authors is deprecated"`                |
| `:location`   | Application-level location of the access   | `#<Thread::Backtrace::Location ...>`     |
| `:backtrace`  | Present if the `:backtrace` option is true | `[ #<Thread::Backtrace::Location ...> ]` |

`:location` and `:backtrace` are computed using the Active Record backtrace
cleaner. In Rails applications, this is the same as `Rails.backtrace_cleaner.

### Active Storage

#### `preview.active_storage`

The event is emitted when a preview is generated for a blob.

For example:

```ruby
blob = ActiveStorage::Blob.find(params[:id])
blob.preview(resize_to_limit: [200, 200]).processed
```

The event payload (`event.payload`) includes the following keys (with typical
example values).

| Payload Key | Description  | Example    |
| ----------- | ------------ | ---------- |

> | `:key`      | Secure token | `"abc123"` |

Reviewer: This is the [key of the blob](https://github.com/rails/rails/blob/cf4e3c508adf930eb70e4b3e6e16d2ca83b7e9b3/activestorage/app/models/active_storage/blob.rb#L166-L167) (I thought at first it might be the key of the service).
Maybe we can make that a bit clearer?
Maybe "Secure token of the blob", "Secure key pointing to the file" or maybe update the example as well:
```suggestion
| `:key`      | Key of the blob | `"secure_token"` |
```

This would apply to the other descriptions of `key` as well.

Author: Makes sense, I've applied it to all the other descriptions of key as well.

#### `transform.active_storage`

The event is emitted when a variant/representation transformation is performed.

For example:

```ruby

> image = current_user.avatar.variant(resize_to_limit: [300, 300]).processed

Reviewer: I don't think we need to assign to an unused variable here:
```suggestion
current_user.avatar.variant(resize_to_limit: [300, 300]).processed
```

example values).

| Payload Key | Description             | Example                                 |
| ----------- | ----------------------- | --------------------------------------- |
| `:key`      | Key used in the store   | `"expensive:calc"`                      |
| `:store`    | Name of the store class | `"ActiveSupport::Cache::MemCacheStore"` |

Cache stores may add their own data as well.

#### `cache_write_multi.active_support`

The event is emitted when multiple values are written to the cache.

For example:

```ruby
Rails.cache.write_multi("a" => 1, "b" => 2)
```

The event payload (`event.payload`) includes the following keys (with typical
example values).

| Payload Key | Description                          | Example                       |
| ----------- | ------------------------------------ | ----------------------------- |
| `:key`      | Keys and values written to the store | `{ "a"=>1, "b"=>2 }`          |
| `:store`    | Name of the store class              | `"ActiveSupport::Cache::..."` |

#### `cache_increment.active_support`

The event is emitted when a counter is incremented in the cache.

For example:

```ruby

> Rails.cache.increment("counter", 5)

Reviewer: It would be nice to have this match the payload table:
```suggestion
Rails.cache.increment("bottles-of-beer", 99)
```

```

The event payload (`event.payload`) includes the following keys (with typical
example values).

| Payload Key | Description             | Example                                   |
| ----------- | ----------------------- | ----------------------------------------- |
| `:key`      | Key used in the store   | `"bottles-of-beer"`                       |
| `:store`    | Name of the store class | `"ActiveSupport::Cache::RedisCacheStore"` |
| `:amount`   | Increment amount        | `99`                                      |

#### `cache_decrement.active_support`

The event is emitted when a counter is decremented in the cache.

For example:

```ruby

> Rails.cache.decrement("counter", 1)

Reviewer: ```suggestion
Rails.cache.decrement("bottles-of-beer", 1)
```

[ActiveSupport::Cache::FileStore]:
    https://api.rubyonrails.org/classes/ActiveSupport/Cache/FileStore.html
[ActiveSupport::Cache::MemCacheStore]:
    https://api.rubyonrails.org/classes/ActiveSupport/Cache/MemCacheStore.html
[ActiveSupport::Cache::MemoryStore]:
    https://api.rubyonrails.org/classes/ActiveSupport/Cache/MemoryStore.html
[ActiveSupport::Cache::RedisCacheStore]:
    https://api.rubyonrails.org/classes/ActiveSupport/Cache/RedisCacheStore.html
[ActiveSupport::Cache::Store#fetch]:
    https://api.rubyonrails.org/classes/ActiveSupport/Cache/Store.html#method-i-fetch
[ActiveSupport::Cache::Store#fetch_multi]:
    https://api.rubyonrails.org/classes/ActiveSupport/Cache/Store.html#method-i-fetch_multi

### Active Support: Messages

#### `message_serializer_fallback.active_support`

The event is emitted when the configured primary message serializer fails and
Active Support falls back to a different serializer.

For example:

```ruby
ActiveSupport::MessageEncryptor.default_message_serializer = :json_allow_marshal
enc = ActiveSupport::MessageEncryptor.new(
  ActiveSupport::KeyGenerator.new("secret").generate_key("salt", 32)

> )

Reviewer: Feel free to ignore, but maybe using a local variable for the key makes the example a bit more readable?
```suggestion
key = ActiveSupport::KeyGenerator.new("secret").generate_key("salt", 32)
enc = ActiveSupport::MessageEncryptor.new(key)
```

Author: agreed!
